<a href="https://colab.research.google.com/github/jomscan/CEIABDTA/blob/main/Analisis_Predictivo_Churnv2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análisis Predictivo: Fuga de Clientes en Telecomunicaciones

Este cuaderno tiene como objetivo explorar un caso real de predicción de fuga de clientes usando un árbol de decisión.

## Objetivo
Reflexionar sobre el uso de modelos predictivos y prescriptivos para tomar decisiones empresariales.

📌 **Nota:** No es necesario programar. Simplemente ejecuta las celdas y reflexiona sobre las preguntas propuestas.

In [1]:
# Cargar librerías
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

# Cargar datos desde GitHub
url = "https://raw.githubusercontent.com/taruntiwarihp/dataSets/master/telecom_churn.csv"
df = pd.read_csv(url)
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

## ¿Qué representan estos datos?
- Cada fila representa un cliente.
- Las columnas muestran información sobre su uso de servicios, gastos y si se dio de baja (`Churn`).

### 💬 Reflexión:
¿Qué variables crees que podrían influir más en que un cliente se dé de baja?

Sin analizar los datos pienso que la variable que mas influye es Customer service calls . Muchas llamadas a servicio al cliente , la mayoria son quejas

In [2]:
# Preprocesamiento mínimo (automático)
df.drop(["State", "Area code", "Phone"], axis=1, inplace=True, errors='ignore')
df["Churn"] = df["Churn"].astype(int)
df = pd.get_dummies(df, drop_first=True)

X = df.drop("Churn", axis=1)
y = df["Churn"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [3]:
# Entrenamiento y predicción con árbol de decisión
model = DecisionTreeClassifier(max_depth=5, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.98      0.96       566
           1       0.88      0.67      0.76       101

    accuracy                           0.94       667
   macro avg       0.91      0.83      0.86       667
weighted avg       0.93      0.94      0.93       667



### 💬 Reflexión:
- ¿El modelo predice bien los clientes que se dan de baja?

No está del todo mal. El modelo es bastante bueno para acertar cuando dice que alguien se va a dar de baja (tiene una precisión de 0.88), o sea, cuando predice una baja, casi siempre acierta. Pero el problema es que se le escapan bastantes: tiene un recall de 0.67, lo que quiere decir que no detecta a 1 de cada 3 personas que realmente se van.

- ¿Qué errores comete? ¿Qué consecuencias tendría para la empresa?

acierta cuando dice que alguien se va, pero deja pasar a muchos que también se van y no los ve venir. Las consecuencias que tendria la empresa es no poder anticipàrse a ese 33% , para poder mejorar sus servicios etc perdiendo muchos clientes sin capacidad de reaccion

- ¿Preferirías minimizar los falsos positivos o los falsos negativos?

Preferiria minimizar los falsos negativos ya que para una empresa la perdida es mayor perdiendo a clientes que se van y no poder anticipase que intentar convencer a clientes que no tienen intencion de irse

## De la predicción a la acción: Análisis prescriptivo

Supón que detectas que clientes con muchas llamadas al servicio de atención tienden a darse de baja.

### 💬 Reflexión:
- ¿Qué acción podrías recomendar a la empresa?

A la empresa le recomendaria que pusiera gran importancia a las llamadas de atencion al cliente ya que es un factor importante , despues analizaria el numero de llamadas para determinar apartir de que numero poner un aviso para que el departamento de fidelizacion le ofresca mejoras personalizadas y compensaria a los trabajadores d ese departamento que consigan retener a clientes con alguna bonificacion de esa manera pondrian mas impetu en hacerlo.

- ¿Cómo podrías personalizar la oferta o el servicio?

se podria personalizar la oferta mejorando la tarifa de las horas que mas consume y subiendo en las que menos para que la empresa tampoco tenga tantas perdidas

Este tipo de análisis ayuda a mejorar la toma de decisiones estratégicas con base en los datos.